src/stream/src/common/log_store_impl/kv_log_store/serde.rs

```diff
@@ -654,6 +657,95 @@ impl<S: StateStoreReadIter> LogStoreRowOpStream<S> {
+            let seq_id = row.get_seq_id();
+            let vnode = row.vnode;
+            let row_op = row.op;
+            let row_read_size = row.size;
+            match row_op {
+                LogStoreOp::Row { .. } | LogStoreOp::Update { .. } => {
+                    progress.insert(vnode, (epoch, seq_id));
+                }
+                _ => {}
+            }
+
+            match row_op {
+                LogStoreOp::Row { op, row, .. } => {
+                    read_info.read_one_row(row_read_size);
+                    ops.push(op);
+                    if let Some(chunk) = data_chunk_builder.append_one_row(row) {
+                        let ops = replace(&mut ops, Vec::with_capacity(chunk_size));
+                        read_info.report(&this.metrics);
+                        yield (
+                            take(&mut progress),
+                            KvLogStoreItem::StreamChunk(StreamChunk::from_parts(ops, chunk)),
+                        );
+                    }
+                }
+                LogStoreOp::Update {
+                    new_value,
+                    old_value,
+                    ..
+                } => {
+                    read_info.read_update(row_read_size);
+                    if !data_chunk_builder.can_append_update() {
+                        let ops = replace(&mut ops, Vec::with_capacity(chunk_size));
+                        let chunk = data_chunk_builder.consume_all().expect("must not be empty");
+                        yield (
+                            take(&mut progress),
+                            KvLogStoreItem::StreamChunk(StreamChunk::from_parts(ops, chunk)),
+                        );
+                    }
+                    ops.extend([Op::UpdateDelete, Op::UpdateInsert]);
+                    assert!(data_chunk_builder.append_one_row(old_value).is_none());
+                    if let Some(chunk) = data_chunk_builder.append_one_row(new_value) {
+                        let ops = replace(&mut ops, Vec::with_capacity(chunk_size));
+                        read_info.report(&this.metrics);
+                        yield (
+                            take(&mut progress),
+                            KvLogStoreItem::StreamChunk(StreamChunk::from_parts(ops, chunk)),
+                        );
+                    }
+                }
+                LogStoreOp::Barrier { is_checkpoint } => {
+                    read_info.read_one_row(row_read_size);
+                    read_info.report(&this.metrics);
+                    if let Some(chunk) = data_chunk_builder.consume_all() {
+                        let ops = replace(&mut ops, Vec::with_capacity(chunk_size));
+                        yield (
+                            take(&mut progress),
+                            KvLogStoreItem::StreamChunk(StreamChunk::from_parts(ops, chunk)),
+                        );
+                    }
+                    yield (
+                        take(&mut progress),
```

Comment:
Since we always `take(&mut progress)` before yielding the stream chunk, here when we yield barrier we will always take an empty progress.

Besides, I think the progress of `Row|Update` and `Barrier` should be handled differently. A `Row|Update` belong to a specific vnode, while `Barrier` infers an aligned progress of a set of vnodes. So we should not maintain and get the progress of `Barrier` here in this method, but may instead include the involved vnodes in `LogStoreOp::Barrier`.

Reply:
Fixed, moved the `vnodes` into `KvLogStoreItem::Barrier`, `LogStoreOp::Barrier`.